src/services/keyword.js

```diff
@@ -46,8 +48,57 @@ angular.module("openshiftCommonServices")
       return filteredObjects;
     };
 
+    // Perform a simple weighted search.
+    // weightedSearch is like filterForKeywords, except each field has a weight
+    // and the result is a sorted array of matches. filterFields is an array of
+    // objects with keys `path` and `weight`.
+    var weightedSearch = function(objects, filterFields, keywords) {
+      if (_.isEmpty(keywords)) {
+        return [];
+      }
+
```

Comment:
Can you give some example code?

Reply:
Outermost: 

```JavaScript 
var weightedSearch = function(objects, filterFields, keywords) {
  if(_.isEmpty(keywords)) {
    return [];
  }
  
  var results = _.reduce(objects, function(result, object) {

    // do scoring work .....
    
    if (score > 0) {
      results.push({
        object: object,
        score: score
      });
    }
    return result;
  }, []);

  // Sort first by score, then by display name for items that have the same score.
  var orderedResult = _.orderBy(results, ['score', displayName], ['desc', 'asc']);
  return _.map(orderedResult, 'object');
  
}
```
Could possibly do the `score` and `matchesKeyword` similarly, but def isn't a huge deal.

Reply:
I guess I'm not clear on the benefit.

Since `_.reduce` can change result to anything, you have to read the code carefully to see that it's just appending values to an array. I feel like `_.each` / `push` is easier to understand.

Reply:
Right, I prob should not have suggested `reduce` over like `_.compact(_.map(objects))`, its too general purpose.

Its prob just the imperative vs declarative discussion... I'd rather not see the "noise" of manually making the new array(s) when we have a number of good functions that do that for us.

That said, is def a nit, and not a huge issue 😄